install/kubernetes/eks-cluster.yaml.tpl

```diff
@@ -0,0 +1,96 @@
+apiVersion: eksctl.io/v1alpha5
+kind: ClusterConfig
+metadata:
+  name: ${CLUSTER_NAME}
+  region: ${REGION}
+  version: "1.27"
+  tags:
+    createdBy: eksctl
+    environment: dev
+    karpenter.sh/discovery: ${CLUSTER_NAME}
+
+karpenter:
+  createServiceAccount: true
+  withSpotInterruptionQueue: true
+  defaultInstanceProfile: "KarpenterNodeInstanceProfile-${CLUSTER_NAME}"
+  version: "v0.29.0"
+
+# if karpenter doesn't suffice: https://github.com/eksctl-io/eksctl/blob/main/examples/23-kubeflow-spot-instance.yaml
+managedNodeGroups:
+  - name: builder-ng
+    privateNetworking: true
+    labels: { role: builders }
+    instanceTypes:
+      - m6a.large
+    volumeSize: 100
+    minSize: 0
+    maxSize: 3
+    desiredCapacity: 2
+    iam:
+      withAddonPolicies:
+        ebs: true
+        imageBuilder: true
+addons:
+  - name: vpc-cni
+    attachPolicyARNs:
+      - arn:aws:iam::aws:policy/AmazonEKS_CNI_Policy
+  - name: kube-proxy
+  - name: aws-ebs-csi-driver
+    wellKnownPolicies:
+      ebsCSIController: true
+  - name: coredns
+
+iamIdentityMappings:
+  - arn: "arn:${AWS_PARTITION}:iam::${AWS_ACCOUNT_ID}:role/KarpenterNodeRole-${CLUSTER_NAME}"
+    username: system:node:{{EC2PrivateDNSName}}
+    groups:
+      - system:bootstrappers
+      - system:nodes
+
+iam:
+  withOIDC: true
+  serviceAccounts:
+    - metadata:
+        name: karpenter
+        namespace: karpenter
+      roleName: ${CLUSTER_NAME}-karpenter
+      attachPolicyARNs:
+        - arn:${AWS_PARTITION}:iam::${AWS_ACCOUNT_ID}:policy/KarpenterControllerPolicy-${CLUSTER_NAME}
+      roleOnly: true
+    - metadata:
+        name: ebs-csi-controller-sa
+        namespace: kube-system
+      wellKnownPolicies:
+        ebsCSIController: true
+    - metadata:
+        name: ${CLUSTER_NAME}
+        namespace: ${CLUSTER_NAME}
+      attachPolicy:
+        Version: "2012-10-17"
+        Statement:
+          - Effect: Allow
+            Action:
+              - "ecr:*"
+            Resource:
+              - "arn:aws:ecr:::${ARTIFACTS_REPO_NAME}"
+          - Effect: Allow
+            Action:
+              - "s3:*"
+              - "s3-object-lambda:*"
+            Resource:
+              - "arn:aws:s3:::${ARTIFACTS_BUCKET_NAME}/*"
+              - "arn:aws:s3:::${ARTIFACTS_BUCKET_NAME}"
+    - metadata:
+        name: aws-manager
+        namespace: ${CLUSTER_NAME}
+      attachPolicy:
+        # https://docs.aws.amazon.com/AmazonS3/latest/userguide/using-presigned-url.html
+        Version: "2012-10-17"
+        Statement:
+          - Effect: Allow
+            Action:
+              - "s3:PutObject"
+              - "s3:GetObject"
+            Resource:
+              - "arn:aws:s3:::${ARTIFACTS_BUCKET_NAME}/*"
+              - "arn:aws:s3:::${ARTIFACTS_BUCKET_NAME}"
```

Comment:
Let's tackle this later in a follow up discussion.

I think the K8s SA and identity mapping should not go into the cluster creation, that should stay a separate step. The main reason being the purpose of cluster creation should be to mimick what an end-user cluster might likely look like.

So the K8s SA + IRSA mapping for controller itself should be a separate install step using pure AWS commands.

Reply:
This is for aws-manager, similar to gcp-manager. I think we had @nstogner on the fence whether or not it's a good idea to have the core controller `*-manager` wihtin the cluster and this is related but falls on the cloud provider API side of that install process. I thought we more or less decided cloud provider API actions mostly fell in the domain of an outside provisioner like tf or eksctl (for now anyway).

I think the concern with allowing the controller to create a SA, its IRSA trust policy binding, and assign it S3 permissions is how permissive we have to allow the controller to be (project-level) as opposed to it's cloud API scope to only be working against service account permissions via modifying trust policy bindings (and the GCP equivalent, workload identity bindings).

Perhaps we split the which entity does what role: the controller handles the ISRA binding but the outside provisioner does everything else (creates the role and the initial policy). Unsure.

Reply:
The first IRSA listed here is definitely not what we decided on earlier in the week and I'll be changing it:

```yaml
    - metadata:
        name: ${CLUSTER_NAME} # formerly would have been hardcoded as container-builder
        namespace: ${CLUSTER_NAME} # should be hardcoded as substratus
      attachPolicy:
        Version: "2012-10-17"
        Statement:
          - Effect: Allow
            Action:
              - "ecr:*"
            Resource:
              - "arn:aws:ecr:::${ARTIFACTS_REPO_NAME}"
```

As I look at this more, I think the mega-provider-identity™️  we should be rolling up into **is** the `aws-manager` role or `gcp-manager` SA.

Reply:
`I think the concern with allowing the controller to create the mega-SA, its IRSA trust policy binding, and assign it S3/ECR permissions is how permissive we have to allow the controller to be (project-level)` --> I think Sam was actually saying the Role should be created in the installation step, but instead of with eksctl, using the aws cli.

Reply:
Spot on Nick. That was exactly what I was trying to say. So someone that already has an existing cluster can easily follow along as well.

Reply:
I get it now! Issue created: https://github.com/substratusai/substratus/issues/178